### DISCUSSION TITLE
Help with establishing communication of 1/4 scan Hub75 panel with A-B-C-D using STM32F103C8T6 (bluepill)

### DISCUSSION THREAD

**Discussion by EduardoYpanaque0** [board707/DMD_STM32#97] [Q&A]
I have an STM32F103C6T8 and I'm trying to establish communication with my P10 RGB panel, which uses 1/4 scan and has a HUB75 interface, using only the A-B-C-D connections. However, I'm having trouble getting any coherent communication. My P10 panel is labeled P10-3535(2727)16*32-4S-V1.1 and L-RS3535MZ-P10RGB. Any assistance or guidance on how to achieve this communication would be greatly appreciated. Thank you!

**Comment by board707** [comment]
Hi
Please run an example `examples/STM32F1/dmd_rgb_pattern_test/dmd_rgb_pattern_test.ino`,  but change the class init (line 44) to
```
DMD_RGB <RGB32x16plainS4_DIRECT, COLOR_4BITS_Packed> dmd(mux_list, DMD_PIN_nOE, DMD_PIN_SCLK, custom_rgbpins, DISPLAYS_ACROSS, DISPLAYS_DOWN, ENABLE_DUAL_BUFFER);
```
If anything will be on the panel, please record the video and show the link in this thread.

**Comment by EduardoYpanaque0** [comment]
The example I'm using is dmd_rgb_pattern_test. I proceeded to change that line of code, but it remains as shown in the following figure (one observation is that when I reset the STM, all LEDs turn off and it stays black). Thank you very much for the prompt response.

![Image1](https://github.com/board707/DMD_STM32/assets/109826518/4eec412e-b2af-47f4-b625-1b19ee32d369)

**Reply by board707** [reply]
How do you upload the code to the STM? Via St-link or Serial uploader?
What the Arduino Ide says after compilation about used memory?

**Reply by EduardoYpanaque0** [reply]
I upload the code to my STM32 using the STM32CubeProgrammer. What I do first is create the binary using the Arduino IDE (where I select the board from here: https://github.com/rogerclarkmelbourne/Arduino_STM32), and then I upload that binary to my STM32 using the STM32CubeProgrammer.

**Reply by board707** [reply]
Could you copy a final Arduino IDE compilation message?

**Reply by EduardoYpanaque0** [reply]
I'm sorry, now I'll share it with you in TXT format.


[ArduinoIDE_compilation.txt](https://github.com/board707/DMD_STM32/files/14890839/ArduinoIDE_compilation.txt)

**Reply by board707** [reply]
In the title of the thread you wrote that your controller is STM32F103**C6T8**, but at compile you select the STM32F103**CBT8** board. These stm32F103 variants (C6 vs CB) are very different. What is your board exactly?

Also, as I see in the log, you upload method is StLink. 
In that case you need to add the line at the very beginning of the `setup()` function:
`enableDebugPorts();`

**Reply by EduardoYpanaque0** [reply]
I apologize for not sending it correctly. Here is my compilation message:

[ArduinoIDE_compilation.txt](https://github.com/board707/DMD_STM32/files/14890995/ArduinoIDE_compilation.txt)

This is a capture of my board:

![Image2](https://github.com/board707/DMD_STM32/assets/109826518/7f07fcbe-f123-467b-be77-cd72476b7db3)

And this is my complete code:


`
/*--------------------------------------------------------------------------------------
 Demo for RGB panels

 DMD_STM32a example code for STM32F103xxx board

 Pattern test for 32x16 1/2 1/4 RGB panels
 ------------------------------------------------------------------------------------- */
 #include "DMD_RGB.h"

//Number of panels in x and y axis
#define DISPLAYS_ACROSS 1
#define DISPLAYS_DOWN 1

// Output buffering - false
#define ENABLE_DUAL_BUFFER false

// ==== DMD_RGB pins ====
// mux pins - A, B, C... all mux pins must be selected from same port!
#define DMD_PIN_A PB6
#define DMD_PIN_B PB5
#define DMD_PIN_C PB4
#define DMD_PIN_D PB3
// put all mux pins at list
uint8_t mux_list[] = { DMD_PIN_A , DMD_PIN_B , DMD_PIN_C , DMD_PIN_D };

// pin OE must be one of PB0 PB1 PA6 PA7
#define DMD_PIN_nOE PB0
#define DMD_PIN_SCLK PB7

// Pins for R0, G0, B0, R1, G1, B1 channels and for clock.
// By default the library uses RGB color order.
// If you need to change this - reorder the R0, G0, B0, R1, G1, B1 pins.
// All this pins also must be selected from same port!
uint8_t custom_rgbpins[] = { PA15, PA0,PA1,PA2,PA3,PA4,PA5 }; // CLK, R0, G0, B0, R1, G1, B1

// Fire up the DMD object as dmd<MATRIX_TYPE, COLOR_DEPTH>

// For pattern testing use one of matrix below:
// RGB32x16plainS4         - 32x16 1/4 Binary mux
// RGB32x16plainS4_DIRECT  - 32x16 1/4 DIRECT mux
// RGB32x16plainS2_DIRECT  - 32x16 1/2 Binary mux

DMD_RGB <RGB32x16plainS4_DIRECT , COLOR_4BITS_Packed> dmd(mux_list, DMD_PIN_nOE, DMD_PIN_SCLK, custom_rgbpins, DISPLAYS_ACROSS, DISPLAYS_DOWN, ENABLE_DUAL_BUFFER);

uint16_t bg = 0;  // background - black
uint16_t fg = 0;  // foreground

void setup(void)
{
    enableDebugPorts();
    dmd.init();
    fg = dmd.Color888(0, 0, 255);
    dmd.setBrightness(50);
}

void loop(void) {
    // fill the matrix with points row by row
    for (int i = 0; i < dmd.height(); i++) {
        for (int j = 0; j < dmd.width(); j++) {
            dmd.drawPixel(j,i, fg);
            delay(30);
        }
    }
    // clear the screen
    dmd.fillScreen(bg);
}
`

Physically, it still looks like the image; the LEDs don't change, and nothing else happens; it remains static.

**Reply by board707** [reply]
Could you look at the panel and show all types the IC on the rear side?

**Reply by board707** [reply]
Additionally - try to replace the standard Arduino_STM32 board package with custom one:
https://github.com/board707/Arduino_STM32/tree/lto_for_c6
Do you have a different STM32 boards - with C8 or CB version of the F103 controller?

**Reply by EduardoYpanaque0** [reply]
First, let me share a recent image of my P10.

![Image3](https://github.com/board707/DMD_STM32/assets/109826518/d2c9a46d-aa7b-4cf1-b4ec-b5e527f2cea6)

On the label, it mentions:
L-RS3535MZ-P10RGB
IC: SM16208/4953/230605

**Reply by EduardoYpanaque0** [reply]
I'm sorry, I made a mistake in the title; my board is the STM32F103C8T6 (Bluepill).

**Reply by board707** [reply]
> my board is the STM32F103C8T6 (Bluepill).

So you should select the "Generic STM32F103C series" board rather than the "... C6/ fake .."
Your chip setup looks compatible.

**Reply by EduardoYpanaque0** [reply]
I've also managed to upload code directly from the Arduino IDE with this board configuration:

![Image4](https://github.com/board707/DMD_STM32/assets/109826518/94928fd4-9fbe-479e-8dfc-cda68645e11e)

This is the compilation message I get:

[ArduinoIDE_compilationBootloader.txt](https://github.com/board707/DMD_STM32/files/14891233/ArduinoIDE_compilationBootloader.txt)

The issue is that when I do it this way, the P10 remains completely black. The code is the same one I shared with you earlier.

By the way, this is the bootloader file I use to upload code from Arduino: [Link to the bootloader file](https://github.com/rogerclarkmelbourne/STM32duino-bootloader/blob/master/binaries/generic-none_bootloader.bin)

**Reply by board707** [reply]
You must to select  the "Generic STM32F103C series" board if your mcu is STM32F103C8T6 (Bluepill). The "Maple Mini" is unsupported by the library.

As about bootloader - I usually use [generic_boot20_pc13.bin](https://github.com/rogerclarkmelbourne/STM32duino-bootloader/blob/master/binaries/generic_boot20_pc13.bin). But using a upload method "STM32CubeProgrammer/ST-LINK" is OK too

**Reply by EduardoYpanaque0** [reply]
I tried again following your recommendation with the [bootloader file](https://github.com/rogerclarkmelbourne/STM32duino-bootloader/blob/master/binaries/generic_boot20_pc13.bin) and then uploaded my program as you indicated.

![Image5](https://github.com/board707/DMD_STM32/assets/109826518/6728c141-a739-46bd-ae93-0e979773e23a)

Now it looks like this, but it doesn't do anything else. That's why I'm sending a photo instead of a video(I changed it to red).

![Image6](https://github.com/board707/DMD_STM32/assets/109826518/2e9c102c-03ae-40fe-873a-c16773339468)

**Reply by board707** [reply]
Where you found this CPU speed - 64 MHz ? What the stm32 package do you used?  The mine doesn't have this value. 
Below are my board settings:
![stm32_board](https://github.com/board707/DMD_STM32/assets/50185434/84be385a-ed33-4128-91b9-cd0f14d776e9)
Please install the package from the link
https://github.com/board707/Arduino_STM32/tree/lto_for_c6
